### PR TITLE
VZ-11268.  Fix scenario where VPO is updated to v1.7.0 and uninstalled before upgrading

### PR DESCRIPTION
**Pull request by michael-cico**
Move isUpgradeRequired() check to the module handlers so that `Module` instances are always created after the VPO is installed in an existing pre-1.7.0 install before the upgrade is issued.

This fixes an edge case where an uninstall is started after VPO is updated from a non-modules (pre-1.7.0) release, but before the system is upgraded.  

Previously Modules for existing installed components pre-1.7.0 were only created once the upgrade was started.  When an uninstall was issued, no `Module` instances existed and uninstall would be broken and/or incomplete.

### CODE REVIEW

File: platform-operator/controllers/verrazzano/controller/reconciler.go [hunk 1]
```diff
@@ -109,20 +109,17 @@ func (r Reconciler) doReconcile(log vzlog.VerrazzanoLogger, controllerCtx contro
 		return res
 	}
 
-	// If an upgrade is pending, do not reconcile; an upgrade is pending if the VPO has been upgraded, but the user
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/controller/reconciler.go:112]: Did you mean to remove the comment

File: platform-operator/controllers/module/component-handler/migration/migration.go [hunk 2]
```diff
@@ -35,6 +38,14 @@ func (h migrationHandler) UpdateStatusIfAlreadyInstalled(ctx handlerspi.HandlerC
 	if err != nil {
 		return result.NewResultShortRequeueDelayWithError(err)
 	}
+
+	if upgradeRequired, err := vzctrlcommon.IsUpgradeRequired(vzcr); upgradeRequired || err != nil {
+		if upgradeRequired {
+			ctx.Log.Oncef("Upgrade required before reconciling module %s", client.ObjectKeyFromObject(module))
+		}
```
Comment (pfmackin) [inline platform-operator/controllers/module/component-handler/migration/migration.go:45]: need to return NewSHortRequeue if no error and upgrade required.  Might be cleaner to remove the 2 ifs

upgradeRequires, err := ...
if err != nil {
  return result.NewResultShortRequeueDelayWithError(err)
}
if upgradeRequired {
  log...
 return result.NewRequeueWithDelay
}

File: platform-operator/controllers/verrazzano/controller/reconciler.go [hunk 3]
```diff
@@ -109,20 +109,17 @@ func (r Reconciler) doReconcile(log vzlog.VerrazzanoLogger, controllerCtx contro
...
-	// has not modified the version in the Verrazzano CR to match the BOM.
-	if upgradePending, err := r.isUpgradeRequired(actualCR); upgradePending || err != nil {
-		controllerCtx.Log.Oncef("Upgrade required before reconciling modules")
-		return result.NewResultShortRequeueDelayIfError(err)
+	if upgradeRequired, err := vzctrlcommon.IsUpgradeRequired(actualCR); upgradeRequired || err != nil {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/controller/reconciler.go:112]: same comment as in controller_utils.go.  Need to return NetShortRequeueWithDelay if no error and upgradRequired

File: platform-operator/controllers/verrazzano/controller/finalizer.go [hunk 4]
```diff
@@ -58,6 +59,21 @@ func (r Reconciler) PreRemoveFinalizer(controllerCtx controllerspi.ReconcileCont
...
+	// user has deleted the VZ CR before initiating an upgrade
+	if !r.doesUninstallConditionExist(actualCR) {
+		effectiveCR, err := r.createEffectiveCR(actualCR)
+		if err != nil {
+			result.NewResultShortRequeueDelayWithError(err)
+		}
+		if result := r.createOrUpdateModules(log, actualCR, effectiveCR); result.ShouldRequeue() {
+			return result
+		}
+		return result.NewResultShortRequeueDelay()
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/controller/finalizer.go:73]: Don't requeue here, just drop through

File: platform-operator/controllers/verrazzano/controller/reconciler.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/controller/reconciler.go:112]: no

File: platform-operator/controllers/module/component-handler/migration/migration.go (hunk 2 above)
Comment (michael-cico) [reply platform-operator/controllers/module/component-handler/migration/migration.go]: done

File: platform-operator/controllers/verrazzano/controller/finalizer.go (hunk 4 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/controller/finalizer.go]: reverted this

File: platform-operator/controllers/verrazzano/controller/reconciler.go (hunk 3 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/controller/reconciler.go]: fixed

File: platform-operator/controllers/verrazzano/controller/reconciler.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/controller/reconciler.go:112]: fixed

**Review by pfmackin — APPROVED** [review]

**Pull request merged**